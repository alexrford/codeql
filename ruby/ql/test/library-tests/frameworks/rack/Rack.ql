private import codeql.ruby.AST
private import codeql.ruby.frameworks.Rack
private import codeql.ruby.DataFlow

query predicate rackApps(Rack::App::App app, DataFlow::ParameterNode env) { env = app.getEnv() }

query predicate rackResponseContentTypes(
  Rack::Response::ResponseNode resp, DataFlow::Node contentType
) {
  contentType = resp.getMimetypeOrContentTypeArg()
}

query predicate redirectResponses(Rack::Response::RedirectResponse resp, DataFlow::Node location) {
  location = resp.getRedirectLocation()
}
