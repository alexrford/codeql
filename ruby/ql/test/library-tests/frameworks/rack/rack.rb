class HelloWorld
  def call(env)
    status = 200
    if something_goes_wrong(env)
      status = 500
    end
    headers = {'Content-Type' => 'text/plain'}
    [status, headers, ['Hello World']]
  end
end

class Proxy
  def initialize(app)
    @app = app
  end

  def call(the_env)
    status, headers, body = @app.call(the_env)
    headers.content_type = "text/html"
    [status, headers, body]
  end
end

class Logger
  def initialize(app, logger = nil)
    @app = app
    @logger = logger
  end

  def call(env)
    began_at = Utils.clock_time
    status, header, body = @app.call(env)
    header = Utils::HeaderHash.new(header)
    body = BodyProxy.new(body) { log(env, status, header, began_at) }
    [status, header, body]
  end
end

class Redirector
  def call(env)
    status = 302
    headers = {'location' => '/foo.html'}
    [status, headers, ['this is a redirect']]
  end
end

class Foo
  def not_call(env)
    [1, 2, 3]
  end
end

class Bar
  def call(env)
    nil
  end
end

class Baz
  def call(env)
    run(env)
  end

  def run(env)
    if env[:foo] == "foo"
      [200, {}, "foo"]
    else
      error
    end
  end

  def error
    [400, {}, "nope"]
  end
end

class Qux
  attr_reader :env
  def self.call(env)
    new(env).call
  end

  def initialize(env)
    @env = env
  end

  def call
    do_redirect
  end

  def do_redirect
    redirect_to = env['redirect_to']
    Rack::Response.new(['redirecting'], 302, 'Location' => redirect_to).finish
  end
end
