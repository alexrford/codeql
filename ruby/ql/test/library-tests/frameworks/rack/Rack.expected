rackApps
| Rack application: -> { ... } | rack_apps.rb:21:17:21:19 | env | rack_apps.rb:21:24:21:48 | call to [] |
| Rack application: Baz | rack.rb:60:12:60:14 | env | rack.rb:66:7:66:22 | call to [] |
| Rack application: Baz | rack.rb:60:12:60:14 | env | rack.rb:73:5:73:21 | call to [] |
| Rack application: ClassApp | rack_apps.rb:16:17:16:19 | env | rack_apps.rb:17:5:17:28 | call to [] |
| Rack application: HelloWorld | rack.rb:2:12:2:14 | env | rack.rb:8:5:8:38 | call to [] |
| Rack application: InstanceApp | rack_apps.rb:6:12:6:14 | env | rack_apps.rb:10:12:10:34 | call to [] |
| Rack application: Logger | rack.rb:30:12:30:14 | env | rack.rb:35:5:35:26 | call to [] |
| Rack application: Proxy | rack.rb:17:12:17:18 | the_env | rack.rb:20:5:20:27 | call to [] |
| Rack application: Qux | rack.rb:79:17:79:19 | env | rack.rb:93:5:93:78 | call to finish |
| Rack application: Redirector | rack.rb:40:12:40:14 | env | rack.rb:43:5:43:45 | call to [] |
| Rack application: { ... } | rack_apps.rb:23:24:23:26 | env | rack_apps.rb:23:29:23:51 | call to [] |
rackResponseContentTypes
| rack.rb:8:5:8:38 | call to [] | rack.rb:7:34:7:45 | "text/plain" |
| rack.rb:20:5:20:27 | call to [] | rack.rb:19:28:19:38 | "text/html" |
redirectResponses
| rack.rb:43:5:43:45 | call to [] | rack.rb:42:30:42:40 | "/foo.html" |
| rack.rb:93:5:93:78 | call to finish | rack.rb:93:60:93:70 | redirect_to |
